src/fsharp/FSharp.Core/iseq.fs

```diff
@@ -0,0 +1,1971 @@
+            inherit TransformFactory<'T,'V>()
+
+            override this.Compose<'W> (outOfBand:IOutOfBand) (pipeIdx:PipeIdx) (next:Activity<'V,'W>) : Activity<'T,'W> =
+                first.Compose outOfBand (pipeIdx-1) (second.Compose outOfBand pipeIdx next)
+
+            static member Combine (first:TransformFactory<'T,'U>) (second:TransformFactory<'U,'V>) : TransformFactory<'T,'V> =
+                upcast ComposedFactory(first, second)
+
+        and IdentityFactory<'T> private () =
+            inherit TransformFactory<'T,'T> ()
+            static let singleton : TransformFactory<'T,'T> = upcast (IdentityFactory<'T>())
+            override __.Compose<'V> (_outOfBand:IOutOfBand) (_pipeIdx:PipeIdx) (next:Activity<'T,'V>) : Activity<'T,'V> = next
+            static member Instance = singleton
+
+        and ISkipable =
+            // Seq.init(Infinite)? lazily uses Current. The only ISeq component that can do that is Skip
+            // and it can only do it at the start of a sequence
+            abstract CanSkip : unit -> bool
+
+        type SeqProcessNextStates =
+        | InProcess  = 0
+        | NotStarted = 1
+        | Finished   = 2
+
+        type Result<'T>() =
+            inherit Folder<'T,'T,NoValue>(Unchecked.defaultof<'T>,Unchecked.defaultof<NoValue>)
+
+            member val SeqState = SeqProcessNextStates.NotStarted with get, set
+
+            override this.ProcessNext (input:'T) : bool =
+                this.Result <- input
+                true
+
+        module Fold =
+            type IIterate<'T> =
+                abstract Iterate<'U,'Result,'State> : outOfBand:Folder<'U,'Result,'State> -> consumer:Activity<'T,'U> -> unit
+
+            [<Struct;NoComparison;NoEquality>]
+            type enumerable<'T> (enumerable:IEnumerable<'T>) =
+                interface IIterate<'T> with
+                    member __.Iterate (outOfBand:Folder<'U,'Result,'State>) (consumer:Activity<'T,'U>) =
+                        use enumerator = enumerable.GetEnumerator ()
+                        let rec iterate () =
+                            if enumerator.MoveNext () then  
+                                consumer.ProcessNext enumerator.Current |> ignore
+                                if outOfBand.HaltedIdx = 0 then
+                                    iterate ()
+                        iterate ()
+
+            [<Struct;NoComparison;NoEquality>]
+            type Array<'T> (array:array<'T>) =
+                interface IIterate<'T> with
+                    member __.Iterate (outOfBand:Folder<'U,'Result,'State>) (consumer:Activity<'T,'U>) =
+                        let array = array
+                        let rec iterate idx =
+                            if idx < array.Length then  
+                                consumer.ProcessNext array.[idx] |> ignore
+                                if outOfBand.HaltedIdx = 0 then
```

Comment:
Hmmmm... I'm trying to remember why the out of band object was used rather than just paramertising ProcessNext... it was fairly early in the evolution of this so I can't honestly remember. I'm pretty sure I considered it - in fact I think that was how I originally coded it; maybe I'll go source code spelunking to see if I can find out...